File: Storage.fs
module Storage

open Tipos
open Unidades


// qtyMMBtu a inyectar (entrada del gasoducto al storage)

let inject  (p: InjectParams) : Operation =
  fun stIn ->
    if stIn.location <> p.storage.location then
      Error (sprintf "Inject: estado en %s, se esperaba %s" stIn.location p.storage.location)
    elif p.qtyIn < 0.0m<MMBTU> then
      Error "Inject: qtyIn < 0"
    elif p.qtyIn > p.storage.injMax then
      Error (sprintf "Inject: qtyIn excede injMax (%f)" ( p.storage.injMax))
    elif stIn.qtyMMBtu < p.qtyIn then
      Error "Inject: qtyMMBtu insuficiente en estado de entrada"
    else
      // Efecto de eficiencia: de qtyIn que entra, solo eff * qtyIn queda como inventario
      let effectiveIn = p.qtyIn * p.storage.injEfficiency
      if p.storage.inv + effectiveIn > p.storage.invMax then
        Error "Inject: invMax excedido"
      else
        // Estado físico del gas: se "parkea" → qtyMMBtu reduce en pipeline
        let stOut = { stIn with qtyMMBtu = stIn.qtyMMBtu - p.qtyIn }
        // Costos: uso por inyección + cargo fijo (si corresponde)
        let cUso =
            let amt = p.qtyIn * p.storage.usageRateInj 
            [{ kind="STORAGE-INJ-USAGE"; qtyMMBtu= p.qtyIn; rate= p.storage.usageRateInj; amount=amt; meta= Map.empty; }]
        let cDem =
          p.storage.demandCharge
          |> Option.map (fun dem -> { kind="STORAGE-DEMAND"; qtyMMBtu=0.0m<MMBTU>; rate= 0m<USD/MMBTU>; amount= dem; meta=Map.empty; })
          |> Option.toList
        let notes =
          [ "storage.invBefore", box (decimal p.storage.inv)
            "storage.effectiveIn", box (decimal effectiveIn)
            "storage.invAfter", box (decimal (p.storage.inv + effectiveIn)) ]
          |> Map.ofList
        Ok { state=stOut; costs=cUso @ cDem; notes=notes }





let withdraw (p: WithdrawParams) : Operation =
  fun stIn ->
    if stIn.location <> p.storage.location then
      Error (sprintf "Withdraw: estado en %s, se esperaba %s" stIn.location p.storage.location)
    elif p.qtyOut < 0.0m<MMBTU> then
      Error "Withdraw: qtyOut < 0"
    elif p.qtyOut > p.storage.wdrMax then
      Error (sprintf "Withdraw: qtyOut excede wdrMax (%f)" p.storage.wdrMax)
    else
      // Eficiencia de retiro: para entregar qtyOut al pipeline, necesitas qtyOut/eff de inventario
      let eff = p.storage.wdrEfficiency
      if eff <= 0.0m then Error "Withdraw: wdrEfficiency <= 0" else
      let invNeeded = p.qtyOut / eff

        
      if invNeeded > p.storage.inv then
        Error "Withdraw: inventario insuficiente"
      else
        let stOut = { stIn with qtyMMBtu = stIn.qtyMMBtu + p.qtyOut }
        let cUso =
            let amt = p.qtyOut * p.storage.usageRateWdr 
            [{ kind="STORAGE-WDR-USAGE"; qtyMMBtu = p.qtyOut; rate = p.storage.usageRateWdr; amount=amt; meta= Map.empty;}]
        let notes =
            [ "storage.invBefore", box (decimal p.storage.inv)
              "storage.invNeeded", box (decimal invNeeded)
              "storage.invAfter",  box (decimal (p.storage.inv - invNeeded)) ]
            |> Map.ofList
              
        Ok { state=stOut; costs=cUso; notes=notes }




let carryCost (p: CarryParams) : Operation =
  fun stIn ->
        // monto = inventario * (apy anual) * (days/365) * (precio de valuación?)
        // Aquí (por simplicidad) consideramos un "costo financiero por unidad" fijo: apy se aplica sobre 1 USD/MMBtu
        // Si deseas valuación mark-to-market, introduce un precio referencial aquí.
        
        let qtyMMBtu = p.storage.inv
        let notionalPerMMBtu = 1m<USD/MMBTU>
        let prorata = (decimal p.days) / 365.0M
        let unitCost = prorata * p.storage.carryAPY * notionalPerMMBtu
        let amount = qtyMMBtu * unitCost 
        let ci =
          { kind="STORAGE-CARRY"
            qtyMMBtu = qtyMMBtu
            rate = unitCost
            amount = amount
            meta= [ "apy", box (float p.storage.carryAPY); "days", box p.days ] |> Map.ofList }
        Ok { state=stIn; costs=[ci]; notes= Map.empty }

File: Operations/Storage.cs
using System;
using System.Collections.Generic;
using ModeloSaavi.Domain;

namespace ModeloSaavi.Operations;

public sealed record StorageState(
    string Location,
    decimal Inventory,
    decimal InventoryMax,
    decimal InjMax,
    decimal WdrMax,
    decimal InjEfficiency,
    decimal WdrEfficiency,
    decimal UsageRateInj,
    decimal UsageRateWdr,
    decimal? DemandCharge,
    decimal CarryApy);

public sealed record InjectParams(StorageState Storage, decimal QtyIn);
public sealed record WithdrawParams(StorageState Storage, decimal QtyOut);
public sealed record CarryParams(StorageState Storage, int Days);

public static class Storage
{
    public static Operation Inject(InjectParams p) => state =>
    {
        if (!string.Equals(state.Location, p.Storage.Location, StringComparison.Ordinal))
            return Result<OpResult>.Fail($"Inject: estado en {state.Location}, se esperaba {p.Storage.Location}");
        if (p.QtyIn < 0) return Result<OpResult>.Fail("Inject: qtyIn < 0");
        if (p.QtyIn > p.Storage.InjMax) return Result<OpResult>.Fail($"Inject: qtyIn excede injMax ({p.Storage.InjMax})");
        if (state.QtyMmbtu < p.QtyIn) return Result<OpResult>.Fail("Inject: qty insuficiente en estado de entrada");

        var effectiveIn = p.QtyIn * p.Storage.InjEfficiency;
        if (p.Storage.Inventory + effectiveIn > p.Storage.InventoryMax)
            return Result<OpResult>.Fail("Inject: invMax excedido");

        var newState = state with { QtyMmbtu = state.QtyMmbtu - p.QtyIn };
        var cUsoAmt = p.QtyIn * p.Storage.UsageRateInj;
        var cUso = new CostItem("STORAGE-INJ-USAGE", p.QtyIn, p.Storage.UsageRateInj, cUsoAmt, new Dictionary<string, object?>());
        var costs = new List<CostItem> { cUso };
        if (p.Storage.DemandCharge is { } dem)
            costs.Add(new CostItem("STORAGE-DEMAND", 0, 0, dem, new Dictionary<string, object?>()));
        var notes = new Dictionary<string, object?>
        {
            ["storage.invBefore"] = p.Storage.Inventory,
            ["storage.effectiveIn"] = effectiveIn,
            ["storage.invAfter"] = p.Storage.Inventory + effectiveIn
        };
        return Result<OpResult>.Ok(new OpResult(newState, costs, notes));
    };

    public static Operation Withdraw(WithdrawParams p) => state =>
    {
        if (!string.Equals(state.Location, p.Storage.Location, StringComparison.Ordinal))
            return Result<OpResult>.Fail($"Withdraw: estado en {state.Location}, se esperaba {p.Storage.Location}");
        if (p.QtyOut < 0) return Result<OpResult>.Fail("Withdraw: qtyOut < 0");
        if (p.QtyOut > p.Storage.WdrMax) return Result<OpResult>.Fail($"Withdraw: qtyOut excede wdrMax ({p.Storage.WdrMax})");
        if (p.Storage.WdrEfficiency <= 0) return Result<OpResult>.Fail("Withdraw: wdrEfficiency <= 0");

        var invNeeded = p.QtyOut / p.Storage.WdrEfficiency;
        if (invNeeded > p.Storage.Inventory) return Result<OpResult>.Fail("Withdraw: inventario insuficiente");

        var newState = state with { QtyMmbtu = state.QtyMmbtu + p.QtyOut };
        var cUsoAmt = p.QtyOut * p.Storage.UsageRateWdr;
        var cUso = new CostItem("STORAGE-WDR-USAGE", p.QtyOut, p.Storage.UsageRateWdr, cUsoAmt, new Dictionary<string, object?>());
        var notes = new Dictionary<string, object?>
        {
            ["storage.invBefore"] = p.Storage.Inventory,
            ["storage.invNeeded"] = invNeeded,
            ["storage.invAfter"] = p.Storage.Inventory - invNeeded
        };
        return Result<OpResult>.Ok(new OpResult(newState, new List<CostItem>{ cUso }, notes));
    };

    public static Operation CarryCost(CarryParams p) => state =>
    {
        var qty = p.Storage.Inventory;
        const decimal notionalPerMmbtu = 1m; // 1 USD/MMBtu referencial
        var prorata = p.Days / 365m;
        var unitCost = prorata * p.Storage.CarryApy * notionalPerMmbtu;
        var amount = qty * unitCost;
        var ci = new CostItem("STORAGE-CARRY", qty, unitCost, amount, new Dictionary<string, object?>
        {
            ["apy"] = p.Storage.CarryApy,
            ["days"] = p.Days
        });
        return Result<OpResult>.Ok(new OpResult(state, new List<CostItem>{ ci }, new Dictionary<string, object?>()));
    };
}
